sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/IncrementalExecution.scala

```diff
@@ -439,6 +439,25 @@ class IncrementalExecution(
               eventTimeWatermarkForEviction = iwEviction)
           ))
 
+      // UpdateEventTimeColumnExec is used to tag the eventTime column, and validate
+      // emitted rows adhere to watermark in the output of transformWithStateInp.
+      // Hence, this node shares the same watermark value as TransformWithStateInPandasExec.
+      // This is the same as above in TransformWithStateExec.
+      // The only difference is TransformWithStateInPandasExec is analysed slightly different
+      // in the physical plan with `Project` and it is not `SerializeFromObjectExec`.
+      case UpdateEventTimeColumnExec(eventTime, delay, None,
+      ProjectExec(projectList, t: TransformWithStateInPandasExec))
```

Comment:
I wonder why Project is necessary between `UpdateEventTimeColumn` and `TransformWithStateInPandas` - could you please check the value of projectList? If that value is the same with output of TransformWithStateInPandasExec, optimizer might be able to remove out Project (in future). If that's the case, we may want to address the pattern of not having a Project in between as well.

Reply:
The output we got from the projectList variable here is the output schema of the output event time column:
> projectList: List(outputTimestamp#15)

I really appreciate you bring the optimization thing up here. After reading your comments, I tried to figure out where this extra projection is coming from and it took me very long to figure out  - we did not add this projection anywhere in any of the event time related change. So then I found this is coming from the pushed down projection from the chaining operator - I chained `groupBy`, but only groupBy on some of the output columns of the TWS output. I am not super familiar with optimization of analyzer, but I think eventually this projection is coming from the pushed down filter from the aggregator operator. So we will see the extra layer of Projection here. That being said, the match case class that contains the `Projection` is not sufficient.
I added another match case class above and I also added another test case to test on cases that matches the situation without the `Projection`.
I did a quick test on Scala side and it doesn't have this issue. It is because TWS node is wrapped by `SerializeFromObjectExec` and this probably does not match with some optimizer rule.

But I think the exact match on the `ProjectExec` is a bit hacky - there is probably more optimization rule I am not aware of. Do you think we should do a recursive match here in case TWS node is (recursively) wrapped by other operators besides `Project`?

Reply:
It's probably easier to not allow pushdown between UpdateEventTimeColumnExec and TransformWithStateExec.

I see filter is not allowed to be pushed down below UpdateEventTimeColumnExec, so it is OK. This might be the case of column pruning - 

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1036-L1045

This could insert the project down to children nodes; we can disallow this for UpdateEventTimeColumnExec, like following:

https://github.com/apache/spark/blob/master/sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/Optimizer.scala#L1031-L1032

Could you give a try with this and see you don't see Projection? If that works, we can just consider UpdateEventTimeColumnExec and TransformWithStateExec to be always close to each other.

Reply:
Thanks for all the code pointers! I added a line in the `Optimizer` and the the column is now not pruned so UpdateEventTimeColumnExec and TransformWithStateInPandasExec are now placed together. I leave the rule inside Optimizer and `IncrementalExecution` will only match with only the case where UpdateEventTimeColumnExec and TransformWithStateInPandasExec are placed together with no extra layer in between.